python/lsst/cp/pipe/bfCoeffCalc.py

```diff
@@ -0,0 +1,1272 @@
+
+        self.debug = lsstDebug.Info(__name__)
+        if self.debug.enabled:
+            self.log.info("Running with debug enabled...")
+            # if we're displaying, test it works, and save the displays for later
+            # it's worth testing here as displays are flaky and sometimes can't be contacted
+            # and given processing takes a while, it's a shame to fail late due to display issues
+            if self.debug.display:
+                try:
+                    afwDisp.setDefaultBackend(self.debug.displayBackend)
+                    afwDisp.Display.delAllDisplays()
+                    self.disp1 = afwDisp.Display(0, open=True)
+                    self.disp2 = afwDisp.Display(1, open=True)
+
+                    im = afwImage.ImageF(1, 1)
+                    im.array[:] = [[1]]
+                    self.disp1.mtv(im)
+                    self.disp1.erase()
+                except NameError:
+                    self.debug.display = False
+                    self.log.warn('Failed to setup/connect to display! Debug display has been disabled')
+
+        plt.interactive(False)  # stop windows popping up when plotting. When headless, use 'agg' backend too
+        self.config.validateIsrConfig(self.log)
+        self.config.validate()
+        self.config.freeze()
+
+    @classmethod
+    def _makeArgumentParser(cls):
+        """Augment argument parser for the BfTask."""
+        parser = pipeBase.ArgumentParser(name=cls._DefaultName)
+        parser.add_argument("--visitPairs", help="The list of visit pairs to use, as a list of tuples "
+                            "enclosed in quotes e.g. \"(123,456),(789,987),(654,321)\""
+                            "NB: must be comma-separated-tuples with no spaces, enclosed in quotes!")
+        parser.add_id_argument("--id", datasetType="bfKernelNew", ContainerClass=BfDataIdContainer,
+                               help="The ccds to use, e.g. --id ccd=0..100")
+        return parser
+
+    @pipeBase.timeMethod
+    def run(self, dataRef, visitPairs):
+        """Run the brighter-fatter measurement task.
+
+        For a dataRef (which is each ccd here), and given a list of visit pairs, calulate the
+        brighter-fatter kernel for the ccd.
+
+        Parameters
+        ----------
+        dataRef : list of lsst.daf.persistence.ButlerDataRef
+            dataRef for the CCD for the visits to be fit.
+        visitPairs : `iterable` of `tuple` of `int`
+            Pairs of visit numbers to be processed together
+        """
+        xcorrs = {}  # dict of lists keyed by either amp or CCD depending on level
+        means = {}
+        kernels = {}
+
+        # setup necessary objects
+        ccdNum = dataRef.dataId[self.config.ccdKey]
+        if self.config.level == 'CCD':
+            xcorrs = {ccdNum: []}
+            means = {ccdNum: []}
+        elif self.config.level == 'AMP':
+            # NB: don't use dataRef.get('raw_detector') due to composites
```

Comment:
I have changed this to read:
```
            # NB: don't use dataRef.get('raw_detector') due to composites
            # because of the way composite objects are handled/constructed (i.e. LSST images)
            # these need to be retrieved from the camera and dereferenced rather than accessed directly
```
Is that a good description? Not really. But the problem is that I don't really know what causes the problem; it's likely a bug, and this is a workaround, but without work digging into some of the messiest bits of the stack it's neither fixable nor possible to give a great description of why this is. Hopefully Gen3 will improve this 😬 🤞

Reply:
I've done something a little different, but it now reads well, thanks for that.